### ISSUE TITLE
Evaluate switching from non-null by default to nullable by default

### ISSUE THREAD

**Issue by milessabin** [gemini-hlsw/gsp-graphql#33]
The GraphQL spec defines types as nullable as default and provides a [Non-Null "wrapping type"](https://graphql.github.io/graphql-spec/June2018/#sec-Type-System.Non-Null) to assert that a field is not nullable.

Early on I decided that for the internals of the compiler and interpreter it would be better to flip this around and treat types as non-nullable by default and add an explicit `NullableType` wrapper, ie. something analogous to `Option`.

This has worked out well in the compiler and interpreter, but was a bit awkward in the introspection component because we need to jump through a few hoops to get the types back into the form that introspection has to deliver back to clients.

This isn't a huge deal, but it gives me pause, because to the extent that people come to Grackle internals with prior GraphQL experience there will be a potentially off-putting mismatch between their expectations and what they'll find. It would at least be good to get a sense for whether this pain is worth the gain.

**Comment by tpolecat** [comment]
I observed the same thing in my earlier attempt at this. Worth keeping an eye on.

**Comment by milessabin** [comment]
I put some time into a spike to see how this would work out and concluded that switching to types being nullable by default and replacing `NullableType` with a new `NonNullType` isn't the right way to go.

The issue is that we represent nullable values internally using `Option` and it simplifies things to have that aligned with a `NullableType` wrapper: as we navigate through types and values, we can simultaneously strip a `NullableType` off the current cursor type and a `Some` off the current cursor focus in a fairly natural way. Flipping that around means we can't rely on common structure between types and terms, instead we need explicit bookkeeping.

This was fiddly enough that although I moved core, circe and generic over I didn't bother taking the experiment all the way through the sql, doobie and skunk modules which are quite a bit more complex.

In conclusion ... worth doing the experiment, but we actually had this right the first time.
